src/expr/src/scalar/func.rs

```diff
@@ -1166,6 +1166,54 @@ fn add_interval<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError>
         .map(Datum::from)
 }
 
+fn bit_not_int16<'a>(a: Datum<'a>) -> Datum<'a> {
```

Comment:
Could the unary funcs be expressed as a [macro](https://github.com/MaterializeInc/materialize/blob/8e8b351490b2bf4cd5e4ecccfa37953c6dee5da4/src/expr/src/scalar/func/macros.rs#L10)? Not sure tbh, but maybe worth checking!

Reply:
Probably, but this issue is more general, as most functions come in families indexed by the possible combination of types (e.g. see `neg_int{XX}`, `abs_int{XX}`, ...). I agree that we should think how we can minimize code for this functions, but I prefer to not tackle this as part of my starter issue. We could use a macro, think about how to utilize generics in a better way, or choose another zero cost abstraction.

I'm happy to open a follow-up issue if more people agree that this is a problem.

Reply:
@petrosagg ^^^

Reply:
I was thinking about something like this:
```
sqlfunc!(
    fn bit_not_int16(a: i16) -> i16 {
        !a
    }
);
```

Elsewhere, there seems to be a `TryFrom` missing:
```
impl TryFrom<Datum<'_>> for i16 {
    type Error = ();
    fn try_from(from: Datum<'_>) -> Result<Self, Self::Error> {
        match from {
            Datum::Int16(f) => Ok(f),
            _ => Err(()),
        }
    }
}

impl TryFrom<Datum<'_>> for Option<i16> {
    type Error = ();
    fn try_from(from: Datum<'_>) -> Result<Self, Self::Error> {
        match from {
            Datum::Null => Ok(None),
            Datum::Int16(f) => Ok(Some(f)),
            _ => Err(()),
        }
    }
}
```

But, totally up to you :D

Reply:
Using the `sqlfunc!` macro for this unary function is totally doable and totally optional since this is your first issue. We're in the process of transitioning the traditional declarations (like the one you provided for `bit_not_int16`) to use `sqlfunc!`. The plan is to transition all functions (even binary and variadic), but for now only unarfy functions are supported by the macro.

So, if you want to merge as is, that's totally fine, I will include this function in a future PR that will migrate all the `int16` functions whole sale.

If you want to dip your toes in the new way of defining unary functions you can use @antiguru's snippets above and a commit like this one for reference: https://github.com/MaterializeInc/materialize/pull/8271/commits/88a144309dc71cca8c350e18570bd9f7fdc6f2b3

Reply:
> So, if you want to merge as is, that's totally fine,  I will include this function in a future PR that will migrate all the int16 functions whole sale.

@petrosagg: I like the approach outlined in #7704! Unfortunately, missed that and started working on a this feature shortly before the PR was merged, so I prefer to merge the current PR without further. I am happy take off some of the refactoring work (including the `~` operator) in the next days.

> Elsewhere, there seems to be a TryFrom missing:

@antiguru Is the above somehow affecting the correctness of the current implementation and actionable on my end?

Reply:
No, `TryFrom` is only needed if you go down the `sqlfunc!` path. Don't worry, I'm slowly converting all the functions in batches, let's merge this as is :)